fs/operations/operations.go

```diff
@@ -830,59 +830,144 @@ func ListLong(ctx context.Context, f fs.Fs, w io.Writer) error {
-	return HashLister(ctx, hash.MD5, f, w)
-}
+// func Md5sum(ctx context.Context, f fs.Fs, w io.Writer) error {
+// 	return HashLister(ctx, hash.MD5, false, false, f, w)
+// }
 
 // Sha1sum list the Fs to the supplied writer
 //
 // Obeys includes and excludes
 //
 // Lists in parallel which may get them out of order
-func Sha1sum(ctx context.Context, f fs.Fs, w io.Writer) error {
-	return HashLister(ctx, hash.SHA1, f, w)
-}
+// func Sha1sum(ctx context.Context, f fs.Fs, w io.Writer) error {
+// 	return HashLister(ctx, hash.SHA1, false, false, f, w)
+// }
 
 // hashSum returns the human readable hash for ht passed in.  This may
 // be UNSUPPORTED or ERROR. If it isn't returning a valid hash it will
 // return an error.
-func hashSum(ctx context.Context, ht hash.Type, o fs.Object) (string, error) {
+func hashSum(ctx context.Context, ht hash.Type, downloadFlag bool, o fs.Object) (string, error) {
+	var sum string
 	var err error
-	tr := accounting.Stats(ctx).NewCheckingTransfer(o)
-	defer func() {
-		tr.Done(ctx, err)
-	}()
-	sum, err := o.Hash(ctx, ht)
-	if err == hash.ErrUnsupported {
-		sum = "UNSUPPORTED"
-	} else if err != nil {
-		fs.Debugf(o, "Failed to read %v: %v", ht, err)
-		sum = "ERROR"
+
+	// If downloadFlag is true, download and hash the file.
+	// If downloadFlag is false, call o.Hash asking the backend for the hash
+	if downloadFlag {
+		
+		// Setup accounting as NewTransfer.  Same approach used in Cat
+		tr := accounting.Stats(ctx).NewTransfer(o)
+		defer func() {
+			tr.Done(ctx, err)
+		}()
+
+		// Setup download options.  Similar approach to Cat.
+		// We do not have a start / end, so we only use fs.Config.DownloadHeaders.
+		// Unsure if copying to a new OpenOption object is required.
+		// Copy also has a very similar setup but includes hashOption
+		var options []fs.OpenOption
+		for _, option := range fs.Config.DownloadHeaders {
+			options = append(options, option)
+		}
+
+		// Using NewReOpen.  Error handling seems to be done different everwhere.  This version is from Cat
+		in, err := NewReOpen(ctx, o, fs.Config.LowLevelRetries, options...)
+		if err != nil {
+			err = fs.CountError(err)
+			fs.Errorf(o, "Failed to open: %v", err)
+			return "ERROR", err
+		}
+		in = tr.Account(ctx, in).WithBuffer() // account and buffer the transfer
+
+		// Unclear if we need to close in.
+		// Copy, which uses NewReOpen does.
+		// Cat, rcat, and maybe some other methods use their own readCloser but don't appear to close
+		// If the error does not occur until close, we log the error but return the hash as okay
+		defer func() {
+			err = in.Close()
+			if err != nil {
+				err = fs.CountError(err)
+				fs.Errorf(o, "Failed to close: %v", err)
+			}
+		}()
+
+		// Transfer is now setup.  Create a multihasher to receive it
+		// Should we instead use hash.StreamTypes here instead of creating the hasher and copying?
+		// We don't fs.CountError or fs.Errorf here because other functions that creat MultiHashers do not.  Should we?
```

Comment:
Confirm direct use of multihasher or use of hash.StreamTypes.
Confirm error handling

Reply:
What you've written looks fine